server/src/main/java/org/opensearch/tasks/TaskResourceTrackingService.java

```diff
@@ -0,0 +1,234 @@
+            resourceAwareTasks.get(taskId).startThreadResourceTracking(threadId, WORKER_STATS, getResourceUsageMetricsForThread(threadId));
+        }
+    }
+
+    /**
+     * Called when a thread finishes working on a task's runnable.
+     *
+     * @param taskId   of the task for which runnable is complete
+     * @param threadId of the thread which executed the runnable and we need to check resource usage for this thread
+     */
+    @Override
+    public void taskExecutionFinishedOnThread(long taskId, long threadId) {
+        if (resourceAwareTasks.containsKey(taskId)) {
+            resourceAwareTasks.get(taskId).stopThreadResourceTracking(threadId, WORKER_STATS, getResourceUsageMetricsForThread(threadId));
+        }
+    }
+
+    public Map<Long, Task> getResourceAwareTasks() {
+        return Collections.unmodifiableMap(resourceAwareTasks);
+    }
+
+    private ResourceUsageMetric[] getResourceUsageMetricsForThread(long threadId) {
+        ResourceUsageMetric currentMemoryUsage = new ResourceUsageMetric(
+            ResourceStats.MEMORY,
+            threadMXBean.getThreadAllocatedBytes(threadId)
+        );
+        ResourceUsageMetric currentCPUUsage = new ResourceUsageMetric(ResourceStats.CPU, threadMXBean.getThreadCpuTime(threadId));
+        return new ResourceUsageMetric[] { currentMemoryUsage, currentCPUUsage };
+    }
+
+    private boolean isThreadWorkingOnTask(Task task, long threadId) {
+        List<ThreadResourceInfo> threadResourceInfos = task.getResourceStats().getOrDefault(threadId, Collections.emptyList());
+
+        for (ThreadResourceInfo threadResourceInfo : threadResourceInfos) {
+            if (threadResourceInfo.isActive()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean validateNoActiveThread(Task task) {
+        for (List<ThreadResourceInfo> threadResourceInfos : task.getResourceStats().values()) {
+            for (ThreadResourceInfo threadResourceInfo : threadResourceInfos) {
+                if (threadResourceInfo.isActive()) return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Adds Task Id in the ThreadContext.
+     * <p>
+     * Stashes the existing ThreadContext and preserves all the existing ThreadContext's data in the new ThreadContext
+     * as well.
+     *
+     * @param task for which Task Id needs to be added in ThreadContext.
+     * @return StoredContext reference to restore the ThreadContext from which we created a new one.
+     * Caller can call context.restore() to get the existing ThreadContext back.
+     */
+    private ThreadContext.StoredContext addTaskIdToThreadContext(Task task) {
+        ThreadContext threadContext = threadPool.getThreadContext();
+
+        boolean staleIdPresentInThreadContext = threadContext.getTransient(TASK_ID) != null
+            && !resourceAwareTasks.containsKey((long) threadContext.getTransient(TASK_ID));
+
```

Comment:
This is now not required because in cases like MultiSearch there could be multiple search requests triggered once the previous finishes. So, the previous one finishes and forks the new search request to new thread (during this the previous request's task id gets copied over as well which we need to handle here. )

Reply:
Glad we caught this :)